File: sdk/go/pkg/artifact/protoc_gen_go.go
package artifact

import (
	"fmt"

	api "github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/api/artifact"
	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/config"
)

func ProtocGenGo(context *config.ConfigContext) (*string, error) {
	name := "protoc-gen-go"
	system := context.GetTargetStr()

	var sourceTarget string

	switch system {
	case "aarch64-darwin":
		sourceTarget = "darwin.arm64"
	case "aarch64-linux":
		sourceTarget = "linux.arm64"
	case "x86_64-darwin":
		sourceTarget = "darwin.amd64"
	case "x86_64-linux":
		sourceTarget = "linux.amd64"
	default:
		return nil, fmt.Errorf("unsupported %s system: %s", name, system)
	}

	sourceVersion := "1.36.11"
	sourcePath := fmt.Sprintf("https://sdk.vorpal.build/source/protoc-gen-go.v%s.%s.tar.gz", sourceVersion, sourceTarget)

	source := NewArtifactSource(name, sourcePath).Build()

	stepScript := `mkdir -p "$VORPAL_OUTPUT/bin"

cp -pr "source/protoc-gen-go/protoc-gen-go" "$VORPAL_OUTPUT/bin/protoc-gen-go"

chmod +x "$VORPAL_OUTPUT/bin/protoc-gen-go"`

	step, err := Shell(context, []*string{}, []string{}, stepScript, []*api.ArtifactStepSecret{})
	if err != nil {
		return nil, err
	}

	steps := []*api.ArtifactStep{step}

	return NewArtifact(name, steps, config.SYSTEMS).
		WithAliases([]string{fmt.Sprintf("%s:%s", name, sourceVersion)}).
		WithSources([]*api.ArtifactSource{&source}).
		Build(context)
}

File: sdk/python/src/vorpal_sdk/artifact/protoc_gen_go.py
"""protoc-gen-go artifact (prebuilt binary).

Mirrors ``sdk/typescript/src/artifact/protoc_gen_go.js``.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from vorpal_sdk.api.artifact import artifact_pb2
from vorpal_sdk.artifact import Artifact, ArtifactSource
from vorpal_sdk.step import shell

if TYPE_CHECKING:
    from vorpal_sdk.context import ConfigContext


class ProtocGenGo:
    """Builder for the protoc-gen-go artifact."""

    def build(self, context: ConfigContext) -> str:
        name = "protoc-gen-go"
        system = context.get_system()

        if system == artifact_pb2.AARCH64_DARWIN:
            source_target = "darwin.arm64"
        elif system == artifact_pb2.AARCH64_LINUX:
            source_target = "linux.arm64"
        elif system == artifact_pb2.X8664_DARWIN:
            source_target = "darwin.amd64"
        elif system == artifact_pb2.X8664_LINUX:
            source_target = "linux.amd64"
        else:
            raise ValueError(f"unsupported {name} system: {system}")

        source_version = "1.36.11"
        source_path = (
            f"https://sdk.vorpal.build/source/"
            f"protoc-gen-go.v{source_version}.{source_target}.tar.gz"
        )

        source = ArtifactSource(name, source_path).build()

        step_script = """mkdir -p "$VORPAL_OUTPUT/bin"

cp -pr "source/protoc-gen-go/protoc-gen-go" "$VORPAL_OUTPUT/bin/protoc-gen-go"

chmod +x "$VORPAL_OUTPUT/bin/protoc-gen-go\""""

        steps = [shell(context, [], [], step_script, [])]
        systems = [
            artifact_pb2.AARCH64_DARWIN,
            artifact_pb2.AARCH64_LINUX,
            artifact_pb2.X8664_DARWIN,
            artifact_pb2.X8664_LINUX,
        ]

        return (
            Artifact(name, steps, systems)
            .with_aliases([f"{name}:{source_version}"])
            .with_sources([source])
            .build(context)
        )
